src/ImageCarouselReact/widget/ImageCarouselReact.tsx

```diff
@@ -0,0 +1,355 @@
+    private descriptionAttr: string;
+    private controls: boolean;
+    private indicators: boolean;
+    private interval: number;
+    private pauseOnHover: boolean;
+    private openPage: string;
+    private slide: boolean;
+    private imageClickMicroflow: string;
+    private staticImageCollection: any[];
+    private width: number;
+    private height: number;
+    private widthUnits: "auto" | "pixels" | "percent" | "viewPort";
+    private heightUnits: "auto" | "pixels" | "percent" | "viewPort";
+    private onClickEvent: "non" | "microflow" | "content" | "popup" | "modal";
+    // Internal variables. Non-primitives created in the prototype are shared between all widget instances.
+    private contextObj: mendix.lib.MxObject;
+    private handles: number[];
+    private data: Idata[];
+    private isLoading: boolean;
+    /**
+     * The TypeScript Contructor, not the dojo consctuctor,
+     * move contructor work into widget prototype at bottom of the page. 
+     */
+    constructor(args?: Object, elem?: HTMLElement) {
+        // Do not add any default value here... it wil not run in dojo!     
+        super() ;
+        return new dojoImageCarouselReact(args, elem);
+    }
+    /**
+     * Create properities for react component based on modeler configurations
+     */
+    public createProps(): ImageCarouselProps {
+        let contextId = this.contextObj ? this.contextObj.getGuid() : "";
+        return {
+            captionAttr: this.captionAttr,
+            contextId,
+            controls: this.controls,
+            data: this.data,
+            dataSourceMicroflow: this.dataSourceMicroflow,
+            descriptionAttr: this.descriptionAttr,
+            entityConstraint: this.entityConstraint,
+            height: this.height,
+            heightUnits: this.heightUnits,
+            imageClickMicroflow: this.imageClickMicroflow,
+            imageEntity: this.imageEntity,
+            imageSource: this.imageSource,
+            indicators: this.indicators,
+            interval: this.interval,
+            isLoading: this.isLoading,
+            onClickEvent: this.onClickEvent,
+            openPage: this.openPage,
+            // openPageModal: this.location, // TODO add to interface later
+            pauseOnHover: this.pauseOnHover,
+            requiresContext: this.requiresContext,
+            slide: this.slide,
+            staticImageCollection: this.staticImageCollection,
+            width: this.width,
+            widthUnits: this.widthUnits,
+        };
+    }
+    /**
+     * dijit._WidgetBase.postCreate is called after constructing the widget.
+     * React widget component, passing in all the props
+     */
+    public postCreate() {
+        logger.debug(this.id + ".postCreate");
+        this.updateRendering();
+    }
+    /**
+     * mxui.widget._WidgetBase.update is called when context is changed or initialized. 
+     *
+     */
+    public update(obj: mendix.lib.MxObject, callback?: Function) {
```

Comment:
@johanflint What are the requirements on the when to call the callback of an **update** ? Should we callback directly, after data retrieval, or after re-rendering the update?